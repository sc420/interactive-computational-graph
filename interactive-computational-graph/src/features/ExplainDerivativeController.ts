import type ChainRuleTerm from "../core/ChainRuleTerm";
import type ExplainDerivativeDescriptionType from "./ExplainDerivativeDescriptionType";
import type ExplainDerivativeItem from "./ExplainDerivativeItem";
import type ExplainDerivativeType from "./ExplainDerivativeType";

type ChainRuleType = "raw" | "previousDerivativesReplaced" | "allReplaced";

const buildExplainDerivativeItems = (
  explainDerivativeType: ExplainDerivativeType,
  nodeId: string,
  nodeDerivative: string,
  targetNodeId: string,
  isReverseMode: boolean,
  chainRuleTerms: ChainRuleTerm[],
): ExplainDerivativeItem[] => {
  switch (explainDerivativeType) {
    case "someValueBecauseChainRule": {
      return buildChainRuleItems(
        nodeId,
        nodeDerivative,
        targetNodeId,
        isReverseMode,
        chainRuleTerms,
      );
    }
    case "oneBecauseFEqualsX": {
      return buildOneBecauseFEqualsXItems(nodeId, targetNodeId, isReverseMode);
    }
    case "zeroBecauseFNotDependsOnX": {
      return buildZeroBecauseFNotDependsOnXItems(
        nodeId,
        targetNodeId,
        isReverseMode,
      );
    }
    case "zeroBecauseXIsConstant": {
      return buildZeroBecauseXIsConstantItems(
        nodeId,
        targetNodeId,
        isReverseMode,
      );
    }
  }
};

const buildOneBecauseFEqualsXItems = (
  nodeId: string,
  targetNodeId: string,
  isReverseMode: boolean,
): ExplainDerivativeItem[] => {
  return [
    buildCalculateDerivativeItem(nodeId, targetNodeId, isReverseMode),
    {
      type: "oneBecauseFEqualsX",
      descriptionParts: [
        {
          type: "text",
          id: "which",
          text: "Which is ",
        },
        {
          type: "latex",
          id: "value",
          latex: "1",
        },
        {
          type: "text",
          id: "because",
          text: " because the derivative target is ",
        },
        {
          type: "latex",
          id: "node",
          latex: nodeId,
        },
        {
          type: "text",
          id: "colon",
          text: ":",
        },
      ],
      latex: "\\displaystyle 1",
    },
  ];
};

const buildZeroBecauseFNotDependsOnXItems = (
  nodeId: string,
  targetNodeId: string,
  isReverseMode: boolean,
): ExplainDerivativeItem[] => {
  return [
    buildCalculateDerivativeItem(nodeId, targetNodeId, isReverseMode),
    {
      type: "zeroBecauseFNotDependsOnX",
      descriptionParts: [
        {
          type: "text",
          id: "which",
          text: "Which is ",
        },
        {
          type: "latex",
          id: "value",
          latex: "0",
        },
        {
          type: "text",
          id: "because",
          text: " because the derivative target doesn't depend on ",
        },
        {
          type: "latex",
          id: "node",
          latex: nodeId,
        },
        {
          type: "text",
          id: "colon",
          text: ":",
        },
      ],
      latex: "\\displaystyle 0",
    },
  ];
};

const buildZeroBecauseXIsConstantItems = (
  nodeId: string,
  targetNodeId: string,
  isReverseMode: boolean,
): ExplainDerivativeItem[] => {
  return [
    buildCalculateDerivativeItem(nodeId, targetNodeId, isReverseMode),
    {
      type: "zeroBecauseXIsConstant",
      descriptionParts: [
        {
          type: "text",
          id: "which",
          text: "Which is ",
        },
        {
          type: "latex",
          id: "value",
          latex: "0",
        },
        {
          type: "text",
          id: "because",
          text: " because ",
        },
        {
          type: "latex",
          id: "node",
          latex: nodeId,
        },
        {
          type: "text",
          id: "is",
          text: " is a constant:",
        },
      ],
      latex: "\\displaystyle 0",
    },
  ];
};

const buildChainRuleItems = (
  nodeId: string,
  nodeDerivative: string,
  targetNodeId: string,
  isReverseMode: boolean,
  chainRuleTerms: ChainRuleTerm[],
): ExplainDerivativeItem[] => {
  return [
    buildCalculateDerivativeItem(nodeId, targetNodeId, isReverseMode),
    {
      type: "useChainRule",
      descriptionParts: [
        {
          type: "text",
          id: "mainText",
          text: "Use the chain rule:",
        },
      ],
      latex: buildChainRuleLatex(
        "raw",
        nodeId,
        targetNodeId,
        isReverseMode,
        chainRuleTerms,
      ),
    },
    {
      type: "previousDerivativesReplaced",
      descriptionParts: buildPreviousDerivativesReplacedDescription(
        targetNodeId,
        isReverseMode,
        chainRuleTerms,
      ),
      latex: buildChainRuleLatex(
        "previousDerivativesReplaced",
        nodeId,
        targetNodeId,
        isReverseMode,
        chainRuleTerms,
      ),
    },
    {
      type: "allReplaced",
      descriptionParts: buildAllReplacedDescription(isReverseMode),
      latex: buildChainRuleLatex(
        "allReplaced",
        nodeId,
        targetNodeId,
        isReverseMode,
        chainRuleTerms,
      ),
    },
    {
      type: "equalTo",
      descriptionParts: [
        {
          type: "text",
          id: "mainText",
          text: "Which is equal to:",
        },
      ],
      latex: `\\displaystyle ${nodeDerivative}`,
    },
  ];
};

const buildCalculateDerivativeItem = (
  nodeId: string,
  targetNodeId: string,
  isReverseMode: boolean,
): ExplainDerivativeItem => {
  return {
    type: "calculatePartialDerivative",
    descriptionParts: [
      {
        type: "text",
        id: "mainText",
        text: "Calculate the partial derivative:",
      },
    ],
    latex: buildPartialDerivativeSubject(nodeId, targetNodeId, isReverseMode),
  };
};

const buildPartialDerivativeSubject = (
  nodeId: string,
  targetNodeId: string,
  isReverseMode: boolean,
): string => {
  const derivative = isReverseMode
    ? buildPartialDerivativeLatex(targetNodeId, nodeId)
    : buildPartialDerivativeLatex(nodeId, targetNodeId);
  return `\\displaystyle ${derivative}`;
};

const buildPreviousDerivativesReplacedDescription = (
  targetNodeId: string,
  isReversedMode: boolean,
  chainRuleTerms: ChainRuleTerm[],
): ExplainDerivativeDescriptionType[] => {
  const side = isReversedMode ? "right" : "left";
  const previousDerivativeTerms: ExplainDerivativeDescriptionType[] =
    chainRuleTerms.map((chainRuleTerm) => {
      let derivative: string;
      if (isReversedMode) {
        derivative = buildPartialDerivativeLatex(
          targetNodeId,
          chainRuleTerm.neighborNodeId,
        );
      } else {
        derivative = buildPartialDerivativeLatex(
          chainRuleTerm.neighborNodeId,
          targetNodeId,
        );
      }
      return {
        type: "latexLink",
        id: `chainRuleTerm-${chainRuleTerm.neighborNodeId}`,
        latex: derivative,
        nodeId: chainRuleTerm.neighborNodeId,
      };
    });

  const commaListParts: ExplainDerivativeDescriptionType[] = [];
  previousDerivativeTerms.forEach((previousDerivativeTerm, index) => {
    if (index > 0) {
      commaListParts.push({
        type: "text",
        id: `part-${index}`,
        text: index === previousDerivativeTerms.length - 1 ? " and " : ", ",
      });
    }
    commaListParts.push(previousDerivativeTerm);
  });

  return [
    {
      type: "text",
      id: "replace",
      text: `Replace ${side}-hand side `,
    },
    ...commaListParts,
    {
      type: "text",
      id: "with",
      text: " with previously calculated value(s):",
    },
  ];
};

const buildAllReplacedDescription = (
  isReversedMode: boolean,
): ExplainDerivativeDescriptionType[] => {
  const side = isReversedMode ? "left" : "right";
  return [
    {
      type: "text",
      id: "mainText",
      text: `Calculate ${side}-hand side derivative(s):`,
    },
  ];
};

const buildChainRuleLatex = (
  chainRuleType: ChainRuleType,
  nodeId: string,
  targetNodeId: string,
  isReverseMode: boolean,
  chainRuleTerms: ChainRuleTerm[],
): string => {
  const multiplicationTerms = buildChainRuleMultiplicationTerms(
    chainRuleType,
    nodeId,
    targetNodeId,
    isReverseMode,
    chainRuleTerms,
  );
  const expression = multiplicationTerms.join("+");
  return `\\displaystyle =${expression}`;
};

const buildPartialDerivativeLatex = (
  numerator: string,
  denominator: string,
): string => {
  return `\\frac{\\partial{${numerator}}}{\\partial{${denominator}}}`;
};

const buildChainRuleMultiplicationTerms = (
  chainRuleType: ChainRuleType,
  nodeId: string,
  targetNodeId: string,
  isReverseMode: boolean,
  chainRuleTerms: ChainRuleTerm[],
): string[] => {
  return chainRuleTerms.map((chainRuleTerm) => {
    let derivative1: string;
    let derivative2: string;
    if (isReverseMode) {
      derivative1 =
        chainRuleType === "allReplaced"
          ? chainRuleTerm.derivativeRegardingCurrent
          : buildPartialDerivativeLatex(chainRuleTerm.neighborNodeId, nodeId);
      derivative2 =
        chainRuleType === "raw"
          ? buildPartialDerivativeLatex(
              targetNodeId,
              chainRuleTerm.neighborNodeId,
            )
          : chainRuleTerm.derivativeRegardingTarget;
    } else {
      derivative1 =
        chainRuleType === "raw"
          ? buildPartialDerivativeLatex(
              chainRuleTerm.neighborNodeId,
              targetNodeId,
            )
          : chainRuleTerm.derivativeRegardingTarget;
      derivative2 =
        chainRuleType === "allReplaced"
          ? chainRuleTerm.derivativeRegardingCurrent
          : buildPartialDerivativeLatex(nodeId, chainRuleTerm.neighborNodeId);
    }
    return `${derivative1} \\cdot ${derivative2}`;
  });
};

export { buildExplainDerivativeItems };
