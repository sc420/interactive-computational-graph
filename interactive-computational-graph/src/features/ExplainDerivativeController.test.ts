import type ChainRuleTerm from "../core/ChainRuleTerm";
import { buildExplainDerivativeItems } from "./ExplainDerivativeController";

test("should build chain rule items of reverse mode", () => {
  const chainRuleTerms: ChainRuleTerm[] = [
    {
      neighborNodeId: "op1",
      derivativeRegardingTarget: "2",
      derivativeRegardingCurrent: "1",
    },
    {
      neighborNodeId: "op2",
      derivativeRegardingTarget: "4",
      derivativeRegardingCurrent: "3",
    },
  ];
  const items = buildExplainDerivativeItems(
    "someValueBecauseChainRule",
    "v1",
    "14",
    "REVERSE",
    "f",
    chainRuleTerms,
  );

  expect(items).toMatchSnapshot();
});

test("should build chain rule items of forward mode", () => {
  const chainRuleTerms: ChainRuleTerm[] = [
    {
      neighborNodeId: "op1",
      derivativeRegardingTarget: "1",
      derivativeRegardingCurrent: "2",
    },
    {
      neighborNodeId: "op2",
      derivativeRegardingTarget: "3",
      derivativeRegardingCurrent: "4",
    },
  ];
  const items = buildExplainDerivativeItems(
    "someValueBecauseChainRule",
    "v1",
    "14",
    "FORWARD",
    "f",
    chainRuleTerms,
  );

  expect(items).toMatchSnapshot();
});

test("should build items when f equals x", () => {
  const chainRuleTerms: ChainRuleTerm[] = [];
  const items = buildExplainDerivativeItems(
    "oneBecauseFEqualsX",
    "x",
    "1",
    "FORWARD",
    "x",
    chainRuleTerms,
  );

  expect(items).toMatchSnapshot();
});

test("should build items when f doesn't depend on x", () => {
  const chainRuleTerms: ChainRuleTerm[] = [];
  const items = buildExplainDerivativeItems(
    "zeroBecauseFNotDependsOnX",
    "x",
    "0",
    "FORWARD",
    "f",
    chainRuleTerms,
  );

  expect(items).toMatchSnapshot();
});

test("should build items when c is a constant", () => {
  const chainRuleTerms: ChainRuleTerm[] = [];
  const items = buildExplainDerivativeItems(
    "zeroBecauseXIsConstant",
    "c",
    "0",
    "FORWARD",
    "f",
    chainRuleTerms,
  );

  expect(items).toMatchSnapshot();
});
