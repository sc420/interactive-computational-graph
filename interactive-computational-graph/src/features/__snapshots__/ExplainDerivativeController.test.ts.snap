// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should build chain rule items of forward mode 1`] = `
Array [
  Object {
    "descriptionParts": Array [
      Object {
        "id": "mainText",
        "text": "Calculate the partial derivative:",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle \\\\frac{\\\\partial{v1}}{\\\\partial{f}}",
    "type": "calculatePartialDerivative",
  },
  Object {
    "descriptionParts": Array [
      Object {
        "id": "mainText",
        "text": "Use the chain rule:",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle =\\\\frac{\\\\partial{op1}}{\\\\partial{f}} \\\\cdot \\\\frac{\\\\partial{v1}}{\\\\partial{op1}}+\\\\frac{\\\\partial{op2}}{\\\\partial{f}} \\\\cdot \\\\frac{\\\\partial{v1}}{\\\\partial{op2}}",
    "type": "useChainRule",
  },
  Object {
    "descriptionParts": Array [
      Object {
        "id": "replace",
        "text": "Replace the left-hand side ",
        "type": "text",
      },
      Object {
        "id": "chainRuleTerm-op1",
        "latex": "\\\\frac{\\\\partial{op1}}{\\\\partial{f}}",
        "nodeId": "op1",
        "type": "latexLink",
      },
      Object {
        "id": "and-1",
        "text": " and ",
        "type": "text",
      },
      Object {
        "id": "chainRuleTerm-op2",
        "latex": "\\\\frac{\\\\partial{op2}}{\\\\partial{f}}",
        "nodeId": "op2",
        "type": "latexLink",
      },
      Object {
        "id": "with",
        "text": " with the previously calculated value(s):",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle =1 \\\\cdot \\\\frac{\\\\partial{v1}}{\\\\partial{op1}}+3 \\\\cdot \\\\frac{\\\\partial{v1}}{\\\\partial{op2}}",
    "type": "previousDerivativesReplaced",
  },
  Object {
    "descriptionParts": Array [
      Object {
        "id": "mainText",
        "text": "Calculate the right-hand side derivative(s):",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle =1 \\\\cdot 2+3 \\\\cdot 4",
    "type": "allReplaced",
  },
  Object {
    "descriptionParts": Array [
      Object {
        "id": "mainText",
        "text": "Which is equal to:",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle 14",
    "type": "equalTo",
  },
]
`;

exports[`should build chain rule items of reverse mode 1`] = `
Array [
  Object {
    "descriptionParts": Array [
      Object {
        "id": "mainText",
        "text": "Calculate the partial derivative:",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle \\\\frac{\\\\partial{f}}{\\\\partial{v1}}",
    "type": "calculatePartialDerivative",
  },
  Object {
    "descriptionParts": Array [
      Object {
        "id": "mainText",
        "text": "Use the chain rule:",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle =\\\\frac{\\\\partial{op1}}{\\\\partial{v1}} \\\\cdot \\\\frac{\\\\partial{f}}{\\\\partial{op1}}+\\\\frac{\\\\partial{op2}}{\\\\partial{v1}} \\\\cdot \\\\frac{\\\\partial{f}}{\\\\partial{op2}}",
    "type": "useChainRule",
  },
  Object {
    "descriptionParts": Array [
      Object {
        "id": "replace",
        "text": "Replace the right-hand side ",
        "type": "text",
      },
      Object {
        "id": "chainRuleTerm-op1",
        "latex": "\\\\frac{\\\\partial{f}}{\\\\partial{op1}}",
        "nodeId": "op1",
        "type": "latexLink",
      },
      Object {
        "id": "and-1",
        "text": " and ",
        "type": "text",
      },
      Object {
        "id": "chainRuleTerm-op2",
        "latex": "\\\\frac{\\\\partial{f}}{\\\\partial{op2}}",
        "nodeId": "op2",
        "type": "latexLink",
      },
      Object {
        "id": "with",
        "text": " with the previously calculated value(s):",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle =\\\\frac{\\\\partial{op1}}{\\\\partial{v1}} \\\\cdot 2+\\\\frac{\\\\partial{op2}}{\\\\partial{v1}} \\\\cdot 4",
    "type": "previousDerivativesReplaced",
  },
  Object {
    "descriptionParts": Array [
      Object {
        "id": "mainText",
        "text": "Calculate the left-hand side derivative(s):",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle =1 \\\\cdot 2+3 \\\\cdot 4",
    "type": "allReplaced",
  },
  Object {
    "descriptionParts": Array [
      Object {
        "id": "mainText",
        "text": "Which is equal to:",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle 14",
    "type": "equalTo",
  },
]
`;

exports[`should build items when c is a constant 1`] = `
Array [
  Object {
    "descriptionParts": Array [
      Object {
        "id": "mainText",
        "text": "Calculate the partial derivative:",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle \\\\frac{\\\\partial{c}}{\\\\partial{f}}",
    "type": "calculatePartialDerivative",
  },
  Object {
    "descriptionParts": Array [
      Object {
        "id": "which",
        "text": "Which is ",
        "type": "text",
      },
      Object {
        "id": "value",
        "latex": "0",
        "type": "latex",
      },
      Object {
        "id": "because",
        "text": " because ",
        "type": "text",
      },
      Object {
        "id": "node",
        "latex": "c",
        "type": "latex",
      },
      Object {
        "id": "is",
        "text": " is a constant:",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle 0",
    "type": "zeroBecauseXIsConstant",
  },
]
`;

exports[`should build items when f doesn't depend on x 1`] = `
Array [
  Object {
    "descriptionParts": Array [
      Object {
        "id": "mainText",
        "text": "Calculate the partial derivative:",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle \\\\frac{\\\\partial{x}}{\\\\partial{f}}",
    "type": "calculatePartialDerivative",
  },
  Object {
    "descriptionParts": Array [
      Object {
        "id": "which",
        "text": "Which is ",
        "type": "text",
      },
      Object {
        "id": "value",
        "latex": "0",
        "type": "latex",
      },
      Object {
        "id": "because",
        "text": " because the derivative target doesn't depend on ",
        "type": "text",
      },
      Object {
        "id": "node",
        "latex": "x",
        "type": "latex",
      },
      Object {
        "id": "colon",
        "text": ":",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle 0",
    "type": "zeroBecauseFNotDependsOnX",
  },
]
`;

exports[`should build items when f equals x 1`] = `
Array [
  Object {
    "descriptionParts": Array [
      Object {
        "id": "mainText",
        "text": "Calculate the partial derivative:",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle \\\\frac{\\\\partial{x}}{\\\\partial{x}}",
    "type": "calculatePartialDerivative",
  },
  Object {
    "descriptionParts": Array [
      Object {
        "id": "which",
        "text": "Which is ",
        "type": "text",
      },
      Object {
        "id": "value",
        "latex": "1",
        "type": "latex",
      },
      Object {
        "id": "because",
        "text": " because the derivative target is ",
        "type": "text",
      },
      Object {
        "id": "node",
        "latex": "x",
        "type": "latex",
      },
      Object {
        "id": "colon",
        "text": ":",
        "type": "text",
      },
    ],
    "latex": "\\\\displaystyle 1",
    "type": "oneBecauseFEqualsX",
  },
]
`;
